Refactor dashboard to use factory and new cache provider

galvani opened the pull request:
~~⚠️ Depends on https://github.com/mautic/mautic/pull/9107 which must be merged first! ⚠️~~

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | N
| Automated tests included? | Y
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR makes the dashboard use the CacheBundle instead of the cache storage.

#### Steps to test this PR:
1. Enable some adapter for the CacheBundle 

For redis I recommend getting some GUI like *Redis Desktop Manager*

2. Pull this PR and clear cache
3. Refresh the /dashboard and verify that data is being written to the cache you have selected
4. Verify cache is being displayed

anton-vlasenko committed:
Add dateFrom and dateTo to widget cache key.
We have to take these parameters into account to generate cache key correctly.

# Conflicts:
#	app/bundles/DashboardBundle/Event/WidgetDetailEvent.php

anton-vlasenko committed:
Add comments and small refactoring.

anton-vlasenko committed:
Fix invalid symbols in cache key by hashing it.

anton-vlasenko committed:
Make test pass (app/bundles/DashboardBundle/Tests/Event/WidgetDetailEventFactoryTest.php)

anton-vlasenko committed:
1. Change DateTime to DateTimeInterface.
2. Move protected methods to the bottom of the file.

# Conflicts:
#	app/bundles/DashboardBundle/Event/WidgetDetailEvent.php

anton-vlasenko committed:
Use exception instead of checking for NULL value.

anton-vlasenko committed:
1. Format the code.
2. Make compatible with php 5.6.

galvani committed:
refactor dashboard to use factory and new cache provider

galvani committed:
revert debug line

galvani committed:
code cleanup

galvani commented:
## Depends on https://github.com/mautic/mautic/pull/9107

galvani committed:
add missing phpdoc

npracht commented:
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

RCheesley commented:
@alanhartless @escopecz another one that needs to be picked up :)

escopecz commented:
This PR depends on https://github.com/mautic/mautic/pull/8817. I can get this one into a good shape once we merge that one.

jordan8037310 commented:
Marking that this is now dependent on #9107

RCheesley commented:
I've assigned this to the 3.2 milestone - will need to get the cache bundle merged in before we can look at merging this PR.

RCheesley commented:
@cla-bot check

anton-vlasenko commented:
This PR now also includes this fix and should also be tested:

[//]: # ( Required: )
#### Description:
Widget content is cached to avoid having to calculate the results each time. The problem is that the cache is used even when the date range is changed and thus the widget still shows the old cached data which is invalid for the new date range.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. open a mautic instance.
3. look at the dashboard
4. change the date range
5. look at the dashboard again and see nothing changed - the date range remains the same
6. clear cache (can be tricky, based on which cache handler is enabled. You can try to use the php console mautic:cache:clear command)
7. see that the widgets now show correct data

#### Steps to test this PR:
1. Load up [this PR]
2. look at the dashboard
3. change the date range
4. look at the dashboard again and see that the date is changed

anton-vlasenko commented:
@cla-bot check

anton-vlasenko committed:
Remove duplicate config.php entry.

anton-vlasenko committed:
Fix the test.

anton-vlasenko committed:
Fix the Dashboard model.

anton-vlasenko committed:
Fix code style.

anton-vlasenko committed:
Fix PHPStan error.

kuzmany commented on app/bundles/DashboardBundle/Config/config.php:
```diff
@@ -77,6 +78,14 @@
+                'arguments' => [
+                    'translator',
+                    'mautic.cache.provider',
+                    'mautic.helper.user',
```
Here could be on third place 'mautic.security' service
I tried push it but do not have permission

anton-vlasenko replied:
@kuzmany I don't have time to fix it now. But I agree with you on that.
I'll try to find time to fix it.

lukassykora committed:
Fix issues from Rector

lukassykora committed:
Fix DashboardModelTest.php

tejasync committed:
Added migrations to fix email_stats:tokens charset to utf8mb4.

kuzmany committed:
Fix unsubscribe form usage

Commit:
Removing check for Alter query

Damzoneuh committed:
add functional test

kuzmany committed:
CS fixer

kuzmany committed:
PHP 7.4 changes to unit tests

kuzmany committed:
Add unit test

kuzmany commented:
@escopecz can you resolve conflicts please?

kuzmany reviewed (changes requested):
From code perspective and from PHPStorm I think the cache bundle is not used


![image](https://user-images.githubusercontent.com/462477/136261517-628c01a2-d6e1-436d-85e0-79849a6255b4.png)


usesLegacyCache is still true.

@escopecz can you check it?

escopecz commented:
@kuzmany you are right. This PR is missing some changes from the orignal PR in our repo. None of the commits has the same commit messages as the orignal PR does and a lot of commits on top of it. Well, 2 years old PR...

escopecz committed:
Adding missing code manually

We cherry-pick commits from our internal PRs to the community. This PR has completely different commits and missing some code changes. So adding those changes manually.

escopecz commented:
I hunted down the missing changes and added them in https://github.com/mautic/mautic/pull/7785/commits/e88b7c863839fdf64c81e12b775ad770c582c8d3. I don't know what happened here.

escopecz committed:
CS fix

escopecz committed:
Adding the filesystem dependency back

as it's used in other method

escopecz committed:
CS fixes

escopecz committed:
Fixing Access to an undefined property

escopecz committed:
Removing unnecessary tab

escopecz committed:
Adding missed dependency in the config

kuzmany commented:
If I import default dashboards, I see these errors:

![image](https://user-images.githubusercontent.com/462477/137104013-5824dc76-c59d-42fc-839e-3125c6bc239f.png)

Commit:
pass in editor, dont rely on this

mollux committed:
remove dependency on pre-installed mautic (local.php file) when running tests

mollux committed:
ensure only sql queries are stored in the fresh_db.sql file

mollux committed:
ensure no side effects and artifacts exist after the testrun

mollux committed:
how does this not fail on github?

mollux committed:
Run all tests in 1 go

mollux committed:
phpstan fixes

mollux committed:
fixed error during mysqldump

mollux committed:
ensure correct error reporting level is used during tests (e.g. to fail on notices and warnings)

mollux committed:
fix deprecation warnings

kuzmany reviewed (changes requested):
I see this after checkout

![image](https://user-images.githubusercontent.com/462477/151833559-fcd5e033-47b5-4086-b016-85ba48d8519d.png)

mollux committed:
switched to pwd via env var instead of insecure in command

mollux committed:
removed leftover code

mollux committed:
tackled column-statistics warning in a more generic way

kuzmany committed:
Add new functional test to check form of unsubscribe page

kuzmany committed:
CS Fixer

mollux committed:
reworked unneeded docblock

kuzmany committed:
Add more functional tests

kuzmany committed:
Merge branch '4.x' into fix-unsubscribe-form-usage

kuzmany committed:
Merge pull request #9752 from kuzmany/fix-unsubscribe-form-usage

kuzmany committed:
Merge branch '4.x' into fix-10455-edit-code-is-broken

kuzmany committed:
Merge pull request #10842 from adiux/fix-10455-edit-code-is-broken

kuzmany committed:
Merge branch '4.x' into improve-local-phpunit-execution

kuzmany committed:
Merge pull request #10759 from mollux/improve-local-phpunit-execution

kuzmany committed:
Merge branch '4.x' into support_for_emoji_in_emails2

kuzmany committed:
Merge pull request #10844 from ts-navghane/support_for_emoji_in_emails2

escopecz commented:
I will have to get to this one some time later. It is not a priority

escopecz committed:
Merge branch '5.x' into pr/galvani/7785

escopecz committed:
Fixing namespaces

escopecz committed:
Fixes caused by the merge

escopecz committed:
Using autowiring, types, PHP8

escopecz committed:
Use constructor for required values

dadarya0 committed:
Merge branch '5.x' into staging.dashboard_cache_improved

dadarya0 committed:
fix phpstan and cs fixer

dadarya0 committed:
Merge remote-tracking branch 'origin' into staging.dashboard_cache_improved

dadarya0 committed:
sync with new code

dadarya0 committed:
Merge branch '5.x' into staging.dashboard_cache_improved

dadarya0 committed:
added service as it was not autowired

dadarya0 committed:
fix rector issue

dadarya0 committed:
fix phpstan issues

dadarya0 committed:
fix phpstan and testcase

dadarya0 committed:
Merge branch '5.x' into staging.dashboard_cache_improved

dadarya0 committed:
increase code coverage

escopecz commented on app/bundles/DashboardBundle/Config/config.php:
```diff
@@ -42,13 +42,22 @@
                     'session',
                 ],
             ],
+            'mautic.dashboard.widget_detail.event.factory' => [
+                'class'     => Mautic\DashboardBundle\Event\WidgetDetailEventFactory::class,
+                'arguments' => [
+                    'translator',
+                    'mautic.cache.provider',
+                    'mautic.security',
+                    'mautic.helper.user',
+                ],
+            ],
```
This can be removed as M5 comes with autowiring
```suggestion
```

dadarya0 replied:
By default we are ignoring autowiring for Event directory so I had added this as service but now I have loaded This factory in autowiring service file.

escopecz replied:
@dadarya0 I see. In that case I'd recommend to move the `WidgetDetailEventFactory` to another folder because we shouldn't be mixing service classes with DTO classes. Can you move it to another folder? Like `app/bundles/DashboardBundle/Factory/WidgetDetailEventFactory.php`?

dadarya0 replied:
Moved to new folder named Factory

escopecz commented on app/bundles/DashboardBundle/Event/WidgetDetailEvent.php:
```diff
@@ -149,26 +175,41 @@ public function getTemplate()
+     * @throws \Psr\Cache\InvalidArgumentException
      */
-    public function setTemplateData(array $templateData, $skipCache = false): void
+    public function setTemplateData(array $templateData, ?bool $skipCache = false): bool
```
We cannot change the method signature. Pease add the return types as method annotations. Is it necessary to change the return type? It's not used anywhere as far as I can see.

If it's necessary to change the types, then please create a new method for it and mark this one as deprecated.

dadarya0 replied:
This function was returning boolean value in return statement but now I have verified that response of this function is not used anywhere so I have reverted again return type to void.

escopecz commented on app/bundles/DashboardBundle/Event/WidgetDetailEvent.php:
```diff
@@ -254,14 +307,6 @@ public function getTranslator()
-    /**
-     * Set security object to check the perimissions.
-     */
-    public function setSecurity(CorePermissions $security): void
```
Please, revert this change as some plugin can call this public method. Mark it as deprecated instead.

escopecz commented on app/bundles/DashboardBundle/Tests/Model/DashboardModelTest.php:
```diff
@@ -6,60 +6,68 @@
-    private \Mautic\DashboardBundle\Model\DashboardModel $model;
+    private DashboardModel $model;
+
+    /**
+     * @var WidgetDetailEventFactory
+     */
+    private $widgetDetailEventFactory;
```
```suggestion
    private WidgetDetailEventFactory $widgetDetailEventFactory;
```

dadarya0 committed:
addressed code reviewer's comment

escopecz commented on app/bundles/DashboardBundle/Event/WidgetDetailEvent.php:
```diff
@@ -149,26 +175,39 @@ public function getTemplate()
+     * @throws \Psr\Cache\InvalidArgumentException
      */
-    public function setTemplateData(array $templateData, $skipCache = false): void
+    public function setTemplateData(array $templateData, ?bool $skipCache = false): void
```
Please add the new param type to a method comment. Changing a method signature is considered a BC break. We'll have to add the real types for M6.

dadarya0 committed:
resolve review comments

escopecz reviewed (approved):
The code looks great and it works as before. When I change the date range to enforce to load the data from database I can see it works:
![Screenshot 2024-01-31 at 13 14 15](https://github.com/mautic/mautic/assets/1235442/86aae481-e31a-414d-9730-9464dc241607)
And when I refresh the data is loaded from cache as expected:
![Screenshot 2024-01-31 at 13 14 32](https://github.com/mautic/mautic/assets/1235442/cc5f58b4-6dca-4589-bb46-ffd793289351)

kuzmany reviewed (approved):
Code looks good 🫰
Seems working properly.

Just wondering use deprecated CacheStorageHelper. I understand why, but good chance to get rid of it on this place

escopecz commented:
@kuzmany good point. We'll remove it in M6.
